src/nspawn/nspawn-mount.c

```diff
@@ -814,9 +815,40 @@ static int mount_bind(const char *dest, CustomMount *m, uid_t uid_shift, uid_t u
         if (m->rm_rf_tmpdir && chown(m->source, uid_shift, uid_shift) < 0)
                 return log_error_errno(errno, "Failed to chown %s: %m", m->source);
 
+        bool unmount = false;
+        if (idmapping != REMOUNT_IDMAPPING_NONE) {
+                /* UID/GIDs of idmapped mounts are always resolved in the caller's user namespace. In other
+                 * words, they're not nested. If we're doing an idmapped mount from a bind mount that's
+                 * already idmapped itself, the old idmap is replaced with the new one. This means that the
+                 * source uid which we put in the idmap userns has to be the uid of mount source in the
+                 * caller's userns *without* any mount idmapping in place. To get that uid, we clone the
+                 * mount source tree and clear any existing idmapping and temporarily mount that tree over
+                 * the mount source before we stat the mount source to figure out the source uid. */
+                _cleanup_close_ int fd_clone = open_tree_attr_fallback(
+                                AT_FDCWD,
+                                m->source,
+                                OPEN_TREE_CLONE|OPEN_TREE_CLOEXEC,
+                                &(struct mount_attr) {
+                                        .attr_clr = MOUNT_ATTR_IDMAP,
+                                });
+                if (fd_clone >= 0) {
+                        if (move_mount(fd_clone, "", AT_FDCWD, m->source, MOVE_MOUNT_F_EMPTY_PATH) < 0)
+                                return log_error_errno(errno, "Failed to mount %s without idmapping: %m", m->source);
```

Comment:
i'd really prefer if we wouldn't attach this to the fs tree here. that's after all making the new stuff work like the old stuff. not the other way around.

given that our baseline includes move_mount() and open_tree() anyway I think it would be to rework this whole function to always clone the mount here, and not attach it except at the final place.